src/native/libs/System.Security.Cryptography.Native.Android/net/dot/android/crypto/PalPbkdf2.java

```diff
@@ -0,0 +1,101 @@
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.SecretKeySpec;
+
+public final class PalPbkdf2 {
+    private static final int ERROR_UNSUPPORTED_ALGORITHM = -1;
+    private static final int SUCCESS = 1;
+
+    public static int pbkdf2OneShot(String algorithmName, byte[] password, byte[] salt, int iterations, byte[] destination)
+        throws ShortBufferException, InvalidKeyException {
+
+        // We do not ever expect a ShortBufferException to ever get thrown since the buffer destination length is always
+        // checked. Let it go through the checked exception and JNI will handle it as a generic failure.
+        // InvalidKeyException should not throw except the the case of an empty key, which we already handle. Let JNI
+        // handle it as a generic failure.
+
+        // We use a custom implementation of PBKDF2 instead of the one provided by the Android platform for two reasons:
+        // The first is that Android only added support for PBKDF2 + SHA-2 family of agorithms in API level 26, and we
+        // need to support SHA-2 prior to that.
+        // The second is that PBEKeySpec only supports char-based passwords, whereas .NET supports arbitrary byte keys.
+
+        assert algorithmName != null;
+        assert password != null;
+        assert salt != null;
+        assert iterations > 0;
+        assert destination != null;
+
+        // The .NET side already validates the hash algorithm name inputs.
+        String javaAlgorithmName = "Hmac" + algorithmName;
+        Mac mac;
+
+        try {
+            mac = Mac.getInstance(javaAlgorithmName);
+        }
+        catch (NoSuchAlgorithmException nsae) {
+            assert false : "The algorithm was checked before, so it should be available.";
+            return ERROR_UNSUPPORTED_ALGORITHM;
+        }
+
+        if (password.length == 0) {
+            // SecretKeySpec does not permit empty keys. Since HMAC just zero extends the key, a single zero byte key is
+            // the same as an empty key.
+            password = new byte[] { 0 };
+        }
+
+        SecretKeySpec key = new SecretKeySpec(password, javaAlgorithmName);
+        mac.init(key);
+
+        // Since this is a one-shot, it should not be possible to exceed the extract limit since the .NET side is
+        // limited to the length of a span (2^31 - 1 bytes). It would only take ~128 million SHA-1 blocks to fill an entire
+        // span, and 128 million fits in a signed 32-bit integer.
+        int blockCounter = 1;
+        int destinationOffset = 0;
+        byte[] blockCounterBuffer = new byte[4]; // Big-endian 32-bit integer
+        byte[] block = new byte[mac.getMacLength()];
+        byte[] u = new byte[block.length];
+
+        while (destinationOffset < destination.length) {
+            assert blockCounter > 0; // Given the comment above, this should never overflow.
+
+            writeBigEndianInt(blockCounter, blockCounterBuffer);
+
+            mac.update(salt);
```

Comment:
On line 78 you call reset even though it's always after a doFinal, but here you don't.  It feels to me like either both or neither should be calling reset?

Reply:
You need to call `reset` if you want to use the Mac after `doFinal`. We don't do it here because we just created the `Mac`. We do it at the beginning of each Mac iteration to get it back in to a working state. This call is an `update`.

Reply:
Doesn't it need the reset, though, between U_l of block 1 and U_0 of block 2?

Reply:
Ahhh. Got it.

Reply:
[Docs say `doFinal` does a reset](https://docs.oracle.com/javase/8/docs/api/javax/crypto/Mac.html#doFinal-byte:A-int-)

> A call to this method resets this Mac object to the state it was in when previously initialized via a call to init(Key) or init(Key, AlgorithmParameterSpec). That is, the object is reset and available to generate another MAC from the same key, if desired, via new calls to update and doFinal. 

So We can just get rid of all the resets.